### PR TITLE
feat: asyncAggregateWithRandomness

### LINKED ISSUE #7197: aggregateWithRandomness performance issue when subscribing to all subnets
### Describe the bug

It takes up to 3 minutes on holesky for `aggregateWithRandomness` when subscribing to all subnets

Screenshot 2024-10-25 at 13 55 45

considering 1 epoch = 6.4 minutes, this takes almost 50% of main thread cpu time this cause performance issue for attached validators

cpu profile shows 51.6% for this time

Screenshot 2024-10-25 at 13 59 56

[holesky_subscribe_all_subnets_network.cpuprofile.zip](https://github.com/user-attachments/files/17517479/holesky_subscribe_all_subnets_network.cpuprofile.zip)
[holesky_subscribe_all_subnets_main.cpuprofile.zip](https://github.com/user-attachments/files/17517482/holesky_subscribe_all_subnets_main.cpuprofile.zip)

### Expected behavior

reduce this time, or consider offload this to native using async/await call
cc @wemeetagain @matthewkeil 

### Steps to reproduce

_No response_

### Additional context

_No response_

### Operating system

Linux

### Lodestar version or commit hash

unstable, v1.22.0

### PR DESCRIPTION
**Pull request by matthewkeil**
**Motivation**

Resolves #7197 

Moves signature checks and aggregation for unaggregated attestations off main thread. Will help clear up thread congestion  when subscribing to all subnets.

### CODE REVIEW

File: packages/beacon-node/src/chain/bls/multithread/jobItem.ts [hunk 1]
```diff
@@ -70,7 +70,7 @@ export function jobItemWorkReq(job: JobQueueItem, metrics: Metrics | null): BlsW
       // and not a problem in the near future
```
Comment (twoeths) [inline packages/beacon-node/src/chain/bls/multithread/jobItem.ts]: stale comment
deployed to `feat2` to gather info

File: packages/beacon-node/src/chain/bls/multithread/jobItem.ts [hunk 2]
```diff
@@ -70,7 +70,7 @@ export function jobItemWorkReq(job: JobQueueItem, metrics: Metrics | null): BlsW
...
       // this is monitored on v1.21.0 https://github.com/ChainSafe/lodestar/pull/6894/files#r1687359225
       const timer = metrics?.blsThreadPool.aggregateWithRandomnessMainThreadDuration.startTimer();
-      const {pk, sig} = aggregateWithRandomness(job.sets.map((set) => ({pk: set.publicKey, sig: set.signature})));
+      const {pk, sig} = await asyncAggregateWithRandomness(job.sets.map((set) => ({pk: set.publicKey, sig: set.signature})));
       timer?.();
```
Comment (twoeths) [inline packages/beacon-node/src/chain/bls/multithread/jobItem.ts]: I'm not sure how to make this metric more useful, if not just remove it

File: packages/beacon-node/src/chain/bls/multithread/jobItem.ts (hunk 1 above)
Comment (matthewkeil) [reply packages/beacon-node/src/chain/bls/multithread/jobItem.ts]: Removed

File: packages/beacon-node/src/chain/bls/multithread/jobItem.ts (hunk 2 above)
Comment (matthewkeil) [reply packages/beacon-node/src/chain/bls/multithread/jobItem.ts]: I updated the metric to reflect libuv time and added one for number of sets total.  Wondering if maybe we should make that one a histogram to see number of sets per job though.  What do you think @twoeths?

Comment (twoeths) [reply packages/beacon-node/src/chain/bls/multithread/jobItem.ts]: I don't see the benefit due to `async/await` nature. If there are multiple async calls right after they are still counted in this metric
maybe we should record length of `job.sets` instead? and the count of `asyncAggregateWithRandomness()` calls would be helpful

**Comment by twoeths** [comment]
there are big improvement of this PR, missed attestations + target were reduced from ~10% to 3-4% on a `1k` subscribing to all subnets

Screenshot 2024-10-29 at 17 35 39

**Review by wemeetagain — DISMISSED** [review]

**Comment by matthewkeil** [comment]
> there are big improvement of this PR, missed attestations + target were reduced from ~10% to 3-4% on a `1k` subscribing to all subnets
> 
> Screenshot 2024-10-29 at 17 35 39
> 
> 

Great news! Can we let the metrics soak a bit more before we merge? Just incase anything wonky shows up

**Comment by wemeetagain** [comment]
Oky, will wait until later today to get more data. I'm hoping to cut a new RC by my end of day today and want to include this.

File: packages/beacon-node/src/chain/bls/multithread/jobItem.ts (hunk 2 above)
Comment (matthewkeil) [reply packages/beacon-node/src/chain/bls/multithread/jobItem.ts]: We already collect that.  I removed them from around the asyncAggregateWithRandomness.

https://github.com/ChainSafe/lodestar/blob/mkeil/async-aggregate-with-randomness/packages/beacon-node/src/chain/bls/multithread/index.ts#L429-L432

**Comment by matthewkeil** [comment]
Everything across the board looks healthier from attestations to scrape duration.  The one thing that did stand out though is mesh peers are not as good.  Not sure if that is just a function of the node still being relatively fresh and that will stabilize over the next several hours or not...  CPU utilization is up a bit but i suppose its just working more efficiently with the multithreading.  

Wondering if we can tune that even further by changing the number of libuv threads.  Can play with that in another PR though.

**Comment by wemeetagain** [comment]
>  The one thing that did stand out though is mesh peers are not as good.

I don't think that is an apples-to-apples comparison. your test is using subscribe-all-subnets and other comparable nodes aren't.

**Review by wemeetagain — APPROVED** [review]

**Comment by twoeths** [comment]
attestation job time decreased significantly since this PR, this is the unstable mainnet node

Screenshot 2024-11-05 at 14 28 26

**Comment by twoeths** [comment]
same to "main to worker" latency

Screenshot 2024-11-06 at 10 02 12

**Comment by twoeths** [comment]
"worker to main" latency on unstable mainnet node was significantly improved as well

Screenshot 2024-11-06 at 10 07 31

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.23.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.23.0) :tada:

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.27.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.27.0) :tada:

**Pull request merged**